django/db/backends/sqlite3/base.py

```diff
@@ -554,19 +554,29 @@ def _sqlite_format_dtdelta(conn, lhs, rhs):
     LHS and RHS can be either:
     - An integer number of microseconds
     - A string representing a datetime
+    - A scalar factor (e.g. float)
     """
-    try:
-        real_lhs = datetime.timedelta(0, 0, lhs) if isinstance(lhs, int) else backend_utils.typecast_timestamp(lhs)
-        real_rhs = datetime.timedelta(0, 0, rhs) if isinstance(rhs, int) else backend_utils.typecast_timestamp(rhs)
-        if conn.strip() == '+':
-            out = real_lhs + real_rhs
+    if conn.strip() in ['+', '-']:
+        try:
+            real_lhs = datetime.timedelta(0, 0, lhs) if isinstance(lhs, int) else backend_utils.typecast_timestamp(lhs)
+            real_rhs = datetime.timedelta(0, 0, rhs) if isinstance(rhs, int) else backend_utils.typecast_timestamp(rhs)
+            if conn.strip() == '+':
+                out = real_lhs + real_rhs
+            else:
+                out = real_lhs - real_rhs
+        except (ValueError, TypeError):
+            return None
+        # typecast_timestamp returns a date or a datetime without timezone.
+        # It will be formatted as "%Y-%m-%d" or "%Y-%m-%d %H:%M:%S[.%f]"
+        return str(out)
+    elif conn.strip() in ['*', '/']:
+        if not isinstance(lhs, (int, float)) or not isinstance(rhs, (int, float)):
+            raise TypeError('Invalid arguments for timedelta multiplication.')
```

Comment:
I added this check because it actually is covered by a tests that would fail if we returned `None` here.

Reply:
After splitting the code as proposed in https://github.com/django/django/pull/14237#discussion_r612320206 this somehow wasn't necessary anymore. I am not sure what happened there.

Reply:
@xi I agree that we should raise an error, but doing this in a user-defined function is not really helpful. I would rather raise `DatabaseError` in `combine_duration_expression()`, e.g.
```diff
diff --git a/django/db/backends/sqlite3/operations.py b/django/db/backends/sqlite3/operations.py
index 90a4241803..ecb7e2edbb 100644
--- a/django/db/backends/sqlite3/operations.py
+++ b/django/db/backends/sqlite3/operations.py
@@ -356,6 +356,10 @@ class DatabaseOperations(BaseDatabaseOperations):
         fn_params = ["'%s'" % connector] + sub_expressions
         if len(fn_params) > 3:
             raise ValueError('Too many params for timedelta operations.')
+        if connector in ['*', '/'] and not (
+            isinstance(fn_params[0], (int, float)) and isinstance(fn_params[1], (int, float))
+        ):
+            raise DatabaseError('No operator matches the given name and argument types.')
         return "django_format_dtdelta(%s)" % ', '.join(fn_params)
 
     def integer_field_range(self, internal_type):
```

Reply:
That is different because at this point those are expressions, not values. I have not yet fully grasped the concept of resolved expressions and output_fields, so I am not entirely sure who to do the equivalent of `isinstance(…, (int, float))` in that case.

Reply:
True, sorry for an undoable request. Folks can always use `sqlite3.enable_callback_tracebacks(True)` to see this message :shrug: .